core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1984,6 +1984,41 @@ object Stream {
       self.reduce(S.combine(_, _))
 
     /**
+      * Repartitions the input with the function `p`. On each step `p` is applied
+      * to the input and all elements but the last of the resulting sequence
+      * are emitted. The last element is then appended to the next input using the
+      * Semigroup `S`.
+      *
+      * @example {{{
+      * scala> import cats.implicits._
+      * scala> Stream("Hel", "l", "o Wor", "ld").repartition(s => Segment.array(s.split(" "))).toList
+      * res0: List[String] = List(Hello, World)
+      * }}}
+      */
+    def repartition(p: O => Segment[O, Unit])(implicit S: Semigroup[O]): Stream[F,O] = {
+      def go(carry: Option[O], s: Stream[F,O]): Pull[F,O,Unit] = {
+        s.pull.uncons.flatMap {
+          case Some((hds, tl)) =>
+            S.combineAllOption(hds.toVector) match {
```

Comment:
Can we implement this without using `toVector` or any of the other strict operations on `Segment`? If `hds` is infinite, the `toVector` call will throw an OOM.

Reply:
Done, thanks! Let me know if you had something different in mind.